packages/@aws-cdk/app-staging-synthesizer/README.md

```diff
@@ -0,0 +1,291 @@
+  // ...
+}
+
+const app = new App({
+  defaultSynthesizer: AppStagingSynthesizer.customFactory({
+    factory: {
+      obtainStagingResources(stack, context) {
+        const app = App.of(stack);
+        if (!App.isApp(app)) {
+          throw new Error(`Stack ${stack.stackName} must be part of an App`);
+        }
+
+        return new CustomStagingStack(app, `CustomStagingStack-${appId}-${context.environmentString}`, { appId: 'my-app-id' }),
+      },
+    },
+    oncePerEnv: true, // by default
+  }),
+});
+```
+
+### Using an Existing Staging Stack
+
+Use `AppStagingSynthesizer.customResources()` to supply an existing stack as the Staging Stack.
+Make sure that the custom stack you provide implements `IStagingResources`.
+
+```ts
+import { IStagingResources } from '@aws-cdk/app-staging-synthesizer';
+
+class CustomStagingStack extends Stack implements IStagingResources {
+  // ...
+}
+
+const resourceApp = new App();
+const resources = new CustomStagingStack(resourceApp, 'CustomStagingStack');
+
+const app = new App({
+  defaultSynthesizer: AppStagingSynthesizer.customResources({
+    resources,
+  }),
+});
+```
+
+## Default Staging Stack
+
+The Default Staging Stack includes all the staging resources necessary for CDK Assets. The below example
+is of a CDK App using the `AppStagingSynthesizer` and creating a file asset for the Lambda Function
+source code. As part of the `DefaultStagingStack`, an S3 bucket and IAM role will be created that will be
+used to upload the asset to S3.
+
+```ts
+import * as path from 'path';
+import * as lambda from 'aws-cdk-lib/aws-lambda';
+
+const app = new App({
+  defaultSynthesizer: new AppStagingSynthesizer.defaultResources({ appId: 'my-app-id' }),
+});
+
+const stack = new Stack(app, 'my-stack');
+
+new lambda.Function(stack, 'lambda', {
+  code: lambda.AssetCode.fromAsset(path.join(__dirname, 'assets')),
+  handler: 'index.handler',
+  runtime: lambda.Runtime.PYTHON_3_9,
+});
+
+app.synth();
+```
+
+### Custom Roles
+
+You can customize some or all of the roles you'd like to use in the synthesizer as well,
+if all you need is to supply custom roles (and not change anything else in the `DefaultStagingStack`):
+
+```ts
+const app = new App({
+  defaultSynthesizer: new AppStagingSynthesizer.defaultResources({
+    appId: 'my-app-id',
+    deploymentRoles: {
+      cloudFormationExecutionRole: BoostrapRole.fromRoleArn('arn:aws:iam::123456789012:role/Execute'),
+      deploymentActionRole: BootstrapRole.fromRoleArn('arn:aws:iam::123456789012:role/Deploy'),
+      lookupRole: BoostrapRole.fromRoleArn('arn:aws:iam::123456789012:role/Lookup'),
+    },
+  }),
+});
+```
+
+Or, you can ask to use the CLI credentials that exist at deploy-time:
+
+```ts
+const app = new App({
+  defaultStackSynthesizer: AppStagingSynthesizer.defaultResources({
+    appId: 'my-app-id',
+    deploymentRoles: {
+      cloudFormationExecutionRole: BootstrapRole.cliCredentials(),
```

Comment:
This line doesn't make sense, really.

I wonder if `deploymentRoles` should accept an enum-like, and maybe call it `DeploymentIdentities` ?